// @gentrace/core/dist/index.js
'use strict';

var v1Api = require('./api/v1-api.js');
var v2Api = require('./api/v2-api.js');
var configuration = require('./configuration.js');
var init = require('./providers/init.js');
var evaluation = require('./providers/evaluation.js');
var testCase = require('./providers/test-case.js');
var dataset = require('./providers/dataset.js');
var pipeline = require('./providers/pipeline.js');
var pipelineMethods = require('./providers/pipeline-methods.js');
var pipelineRun = require('./providers/pipeline-run.js');
var stepRun = require('./providers/step-run.js');
var plugin = require('./providers/plugin.js');
var handler = require('./providers/handler.js');
var files = require('./providers/files.js');
var runTest = require('./providers/run-test.js');
var testResult = require('./providers/test-result.js');
var evaluator = require('./providers/evaluator.js');
var runners = require('./providers/runners.js');



exports.V1Api = v1Api.V1Api;
exports.V1ApiAxiosParamCreator = v1Api.V1ApiAxiosParamCreator;
exports.V1ApiFactory = v1Api.V1ApiFactory;
exports.V1ApiFp = v1Api.V1ApiFp;
exports.V2Api = v2Api.V2Api;
exports.V2ApiAxiosParamCreator = v2Api.V2ApiAxiosParamCreator;
exports.V2ApiFactory = v2Api.V2ApiFactory;
exports.V2ApiFp = v2Api.V2ApiFp;
exports.Configuration = configuration.Configuration;
Object.defineProperty(exports, 'GENTRACE_API_KEY', {
	enumerable: true,
	get: function () { return init.GENTRACE_API_KEY; }
});
Object.defineProperty(exports, 'GENTRACE_BASE_PATH', {
	enumerable: true,
	get: function () { return init.GENTRACE_BASE_PATH; }
});
Object.defineProperty(exports, 'GENTRACE_BRANCH', {
	enumerable: true,
	get: function () { return init.GENTRACE_BRANCH; }
});
Object.defineProperty(exports, 'GENTRACE_COMMIT', {
	enumerable: true,
	get: function () { return init.GENTRACE_COMMIT; }
});
Object.defineProperty(exports, 'GENTRACE_RESULT_NAME', {
	enumerable: true,
	get: function () { return init.GENTRACE_RESULT_NAME; }
});
Object.defineProperty(exports, 'GENTRACE_RUN_NAME', {
	enumerable: true,
	get: function () { return init.GENTRACE_RUN_NAME; }
});
Object.defineProperty(exports, 'GENTRACE_SHOW_CONNECTION_ERRORS', {
	enumerable: true,
	get: function () { return init.GENTRACE_SHOW_CONNECTION_ERRORS; }
});
exports.deinit = init.deinit;
exports.flush = init.flush;
exports.getGentraceApiKey = init.getGentraceApiKey;
exports.getGentraceBasePath = init.getGentraceBasePath;
Object.defineProperty(exports, 'globalGentraceApi', {
	enumerable: true,
	get: function () { return init.globalGentraceApi; }
});
Object.defineProperty(exports, 'globalGentraceApiV2', {
	enumerable: true,
	get: function () { return init.globalGentraceApiV2; }
});
Object.defineProperty(exports, 'globalGentraceConfig', {
	enumerable: true,
	get: function () { return init.globalGentraceConfig; }
});
Object.defineProperty(exports, 'globalRequestBuffer', {
	enumerable: true,
	get: function () { return init.globalRequestBuffer; }
});
exports.init = init.init;
exports.resetGlobalGentraceApi = init.resetGlobalGentraceApi;
exports.bulkCreateEvaluations = evaluation.bulkCreateEvaluations;
exports.getEvaluations = evaluation.getEvaluations;
exports.createTestCase = testCase.createTestCase;
exports.createTestCases = testCase.createTestCases;
exports.deleteTestCase = testCase.deleteTestCase;
exports.getTestCase = testCase.getTestCase;
exports.getTestCases = testCase.getTestCases;
exports.getTestCasesForDataset = testCase.getTestCasesForDataset;
exports.updateTestCase = testCase.updateTestCase;
exports.createDataset = dataset.createDataset;
exports.getDataset = dataset.getDataset;
exports.getDatasets = dataset.getDatasets;
exports.updateDataset = dataset.updateDataset;
exports.Pipeline = pipeline.Pipeline;
exports.getPipelines = pipelineMethods.getPipelines;
exports.PipelineRun = pipelineRun.PipelineRun;
exports.getRun = pipelineRun.getRun;
exports.StepRun = stepRun.StepRun;
exports.GentracePlugin = plugin.GentracePlugin;
exports.isConfig = handler.isConfig;
exports.uploadBuffer = files.uploadBuffer;
exports.uploadFile = files.uploadFile;
exports.runTest = runTest.runTest;
exports.runTestWithDataset = runTest.runTestWithDataset;
exports.constructSubmissionPayloadAdvanced = testResult.constructSubmissionPayloadAdvanced;
exports.constructSubmissionPayloadSimple = testResult.constructSubmissionPayloadSimple;
exports.getTestResult = testResult.getTestResult;
exports.getTestResultStatus = testResult.getTestResultStatus;
exports.getTestResults = testResult.getTestResults;
exports.submitTestResult = testResult.submitTestResult;
exports.updateTestResult = testResult.updateTestResult;
exports.getEvaluators = evaluator.getEvaluators;
exports.getTestRunners = runners.getTestRunners;
exports.submitTestRunners = runners.submitTestRunners;
exports.updateTestResultWithRunners = runners.updateTestResultWithRunners;
